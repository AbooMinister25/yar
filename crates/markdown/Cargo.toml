[package]
name = "yar_markdown"
version = "0.1.0"
license = "MIT"
edition.workspace = true

[dependencies]
chrono = { version = "0.4.41", features = ["serde"] }
color-eyre.workspace = true
minijinja = "2.11.0"
nom = "7.1.3"
pulldown-cmark = { version = "0.13.0", default-features = false, features = ["html"] }
serde = { version = "1.0.219", features = ["derive"] }
smol_str = { version = "0.3.2", features = ["serde"] }
syntect = { version = "5.2.0", default-features = false, features = ["default-fancy"] }
toml = "0.8.22"

[dev-dependencies]
insta = { workspace = true, features = ["yaml"] }

[lints]
workspace = true
