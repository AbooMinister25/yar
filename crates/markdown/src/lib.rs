use std::sync::LazyLock;

use chrono::{DateTime, NaiveDateTime, TimeZone, Utc};
use color_eyre::Result;
use pulldown_cmark::{
    CodeBlockKind, Event, HeadingLevel, Options, Parser, Tag, TagEnd, html::push_html,
};
use serde::{Deserialize, Serialize};
use syntect::{highlighting::ThemeSet, html::highlighted_html_for_string, parsing::SyntaxSet};

static SYNTAX_SET: LazyLock<SyntaxSet> = LazyLock::new(SyntaxSet::load_defaults_newlines);
static THEME_SET: LazyLock<ThemeSet> = LazyLock::new(ThemeSet::load_defaults);

/// The frontmatter metadata for a parsed markdown document.
#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct Frontmatter {
    pub title: String,
    pub tags: Vec<String>,
    pub template: Option<String>,
    pub completed: Option<bool>,
    pub date: Option<String>,
    pub updated: Option<String>,
    pub series: Option<SeriesInfo>,
    pub slug: Option<String>,
    #[serde(default)]
    pub draft: bool,
}

/// Details about a series that a post belongs to, if any.
#[derive(Debug, Deserialize, Serialize, PartialEq, Eq)]
pub struct SeriesInfo {
    pub part: Option<i32>,
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct TOCHeading {
    id: Option<String>,
    text: String,
}

impl TOCHeading {
    fn new(id: Option<String>, text: String) -> Self {
        Self { id, text }
    }
}

/// A parsed markdown document.
#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct Document {
    pub date: DateTime<Utc>,
    pub updated: DateTime<Utc>,
    pub content: String,
    pub toc: Vec<TOCHeading>,
    pub summary: String,
    frontmatter: Frontmatter,
}

#[derive(Debug)]
struct CodeBlock {
    lang: String,
    text: String,
}

impl CodeBlock {
    pub fn new(lang: String) -> Self {
        Self {
            lang,
            text: "".into(),
        }
    }
}

pub fn render_markdown(content: &str) -> Result<Document> {
    let mut options = Options::empty();
    options.insert(Options::ENABLE_TABLES);
    options.insert(Options::ENABLE_FOOTNOTES);
    options.insert(Options::ENABLE_STRIKETHROUGH);
    options.insert(Options::ENABLE_SMART_PUNCTUATION);
    options.insert(Options::ENABLE_YAML_STYLE_METADATA_BLOCKS);
    options.insert(Options::ENABLE_MATH);

    let frontmatter = parse_frontmatter(content)?;

    let mut html_output = String::new();
    let parser = Parser::new_ext(content, options);

    let mut codeblock = None;
    let mut current_heading = None;
    let mut headings = Vec::new();

    let parser = parser.filter_map(|event| match event {
        // TODO: Emit <pre><code> and highlight line by line.
        Event::Start(Tag::CodeBlock(CodeBlockKind::Fenced(lang))) => {
            let lang = lang.trim();
            codeblock = Some(CodeBlock::new(lang.into()));
            None
        }
        Event::End(TagEnd::CodeBlock) => {
            if let Some(cb) = &codeblock {
                let syntax = SYNTAX_SET
                    .find_syntax_by_extension(&cb.lang)
                    .unwrap_or(SYNTAX_SET.find_syntax_plain_text());
                let html = highlighted_html_for_string(
                    &cb.text,
                    &SYNTAX_SET,
                    syntax,
                    &THEME_SET.themes["solarized-dark"],
                )
                .ok()?;

                codeblock = None;
                Some(Event::Html(html.into()))
            } else {
                None
            }
        }
        Event::Start(Tag::Heading {
            level: HeadingLevel::H2,
            ref id,
            ..
        }) => {
            current_heading = Some(TOCHeading::new(
                id.as_ref().map(|c| c.to_string()),
                "".to_string(),
            ));
            Some(event)
        }
        Event::End(TagEnd::Heading(HeadingLevel::H2)) => {
            headings.push(current_heading.take().expect("Heading end before start?"));
            Some(event)
        }
        Event::Text(ref t) => {
            if let Some(cb) = &mut codeblock {
                cb.text.push_str(t);
            } else if let Some(h) = &mut current_heading {
                h.text.push_str(t);
            }
            Some(event)
        }
        Event::Code(ref s)
        | Event::InlineMath(ref s)
        | Event::DisplayMath(ref s)
        | Event::InlineHtml(ref s) => {
            if let Some(h) = &mut current_heading {
                h.text.push_str(s);
            }
            Some(event)
        }
        _ => Some(event),
    });

    push_html(&mut html_output, parser);

    // Extract dates from frontmatter
    let date = frontmatter.date.as_ref().map_or(
        Ok::<DateTime<Utc>, color_eyre::Report>(Utc::now()),
        |d| {
            let parsed = d.parse::<NaiveDateTime>()?;
            Ok(Utc.from_utc_datetime(&parsed))
        },
    )?;

    let updated = frontmatter.updated.as_ref().map_or(
        Ok::<DateTime<Utc>, color_eyre::Report>(date),
        |d| {
            let parsed = d.parse::<NaiveDateTime>()?;
            Ok(Utc.from_utc_datetime(&parsed))
        },
    )?;

    Ok(Document {
        date,
        updated,
        content: html_output,
        toc: headings,
        summary: "".to_string(),
        frontmatter,
    })
}

fn parse_frontmatter(content: &str) -> Result<Frontmatter> {
    let mut opening_delim = false;
    let mut frontmatter_content = String::new();

    for line in content.lines() {
        if line.trim() == "..." {
            if opening_delim {
                break;
            }

            opening_delim = true;
            continue;
        }

        frontmatter_content.push_str(line);
        frontmatter_content.push('\n');
    }

    let frontmatter = toml::from_str(&frontmatter_content)?;
    Ok(frontmatter)
}
