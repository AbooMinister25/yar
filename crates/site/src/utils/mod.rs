use std::path::{Path, PathBuf};

use color_eyre::{Result, eyre::ContextCompat};
use url::Url;

pub mod fs;

/// Build permalink for a site item.
pub fn build_permalink<P: AsRef<Path>, T: AsRef<Path>>(
    path: P,
    out_dir: T,
    url: &Url,
) -> Result<Url> {
    let out = out_dir
        .as_ref()
        .file_name()
        .context("Output directory shouldn't terminate in ..")?;

    let mut components = path
        .as_ref()
        .components()
        .skip_while(|c| c.as_os_str() != out);
    components.next();
    let mut url_ending = components.collect::<PathBuf>();
    if url_ending.ends_with("index.html") {
        url_ending = url_ending
            .parent()
            .context("path doesn't have parent?")?
            .to_path_buf();
    }
    let permalink = url.join(
        url_ending
            .to_str()
            .context("Path should be valid unicode.")?,
    )?;

    Ok(permalink)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_build_permalink() -> Result<()> {
        let path = Path::new("site/index.html");
        let out_dir = Path::new("site/");
        let url = Url::parse("https://example.com/")?;
        insta::assert_yaml_snapshot!(build_permalink(path, out_dir, &url)?);

        let path = Path::new("site/posts/hello-world/index.html");
        let out_dir = Path::new("site/");
        let url = Url::parse("https://example.com/")?;
        insta::assert_yaml_snapshot!(build_permalink(path, out_dir, &url)?);

        let path = Path::new("site/assets/style.css");
        let out_dir = Path::new("site/");
        let url = Url::parse("https://example.com/")?;
        insta::assert_yaml_snapshot!(build_permalink(path, out_dir, &url)?);

        Ok(())
    }
}
