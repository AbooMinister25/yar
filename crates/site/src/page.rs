use std::fmt::Debug;
use std::fs;
use std::path::{Path, PathBuf};
use std::rc::Rc;

use color_eyre::Result;
use color_eyre::eyre::ContextCompat;
use markdown::{Document, MarkdownRenderer};
use minify_html::{Cfg, minify};
use minijinja::{Environment, context};
use serde::{Deserialize, Serialize};
use url::Url;

use crate::utils::build_permalink;
use crate::utils::fs::ensure_directory;

/// A single page in the site.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Page {
    pub path: PathBuf,
    pub source_hash: String,
    pub out_path: PathBuf,
    pub permalink: Url,
    pub document: Document,
}

impl Page {
    #[allow(clippy::too_many_arguments)]
    pub fn new<P: AsRef<Path>, T: AsRef<Path>, Z: AsRef<Path>>(
        path: P,
        content: &str,
        source_hash: String,
        out_dir: T,
        root: Z,
        url: &Url,
        markdown_renderer: &MarkdownRenderer,
        env: &Environment,
    ) -> Result<Self> {
        let document = markdown_renderer.parse_from_string(content, env)?;
        let out_path = out_path(
            &path,
            &out_dir,
            root,
            &document.frontmatter.title,
            document.frontmatter.slug.as_deref(),
        );
        let permalink = build_permalink(&out_path, out_dir, url)?;

        Ok(Self {
            path: path.as_ref().into(),
            out_path,
            source_hash,
            permalink,
            document,
        })
    }

    pub fn render(&self, index: &[Rc<Self>], env: &Environment) -> Result<()> {
        ensure_directory(
            self.out_path
                .parent()
                .context("Path should have a parent")?,
        )?;

        let frontmatter = &self.document.frontmatter;
        let template = frontmatter.template.as_ref().map_or("post.html", |v| v);
        let template = env.get_template(template)?;

        let rendered_html =
            template.render(context! { document => self.document, pages => index})?;
        let cfg = Cfg::new();
        let minified = minify(rendered_html.as_bytes(), &cfg);

        fs::write(&self.out_path, minified)?;

        Ok(())
    }
}

fn out_path<P: AsRef<Path>, T: AsRef<Path>, Z: AsRef<Path>>(
    path: P,
    out_dir: T,
    root: Z,
    title: &str,
    slug: Option<&str>,
) -> PathBuf {
    let out_dir = out_dir.as_ref();

    let ending = if path.as_ref().ends_with("index.md") {
        PathBuf::from("index.html")
    } else {
        PathBuf::from(slug.map_or_else(|| title.replace(' ', "-"), ToOwned::to_owned))
            .join("index.html")
    };

    let mut components = path
        .as_ref()
        .parent()
        .unwrap_or_else(|| path.as_ref())
        .components()
        .filter(|c| !c.as_os_str().to_str().is_some_and(|s| s.starts_with('_')));

    if root.as_ref() != Path::new(".") {
        components.next();
    }

    out_dir
        .components()
        .chain(components)
        .collect::<PathBuf>()
        .join(ending)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_out_path() {
        let path = out_path(
            "site/_content/posts/hello-world.md",
            "public",
            "site",
            "hello world",
            None,
        );
        insta::assert_yaml_snapshot!(path);

        let path = out_path(
            "site/_content/posts/hello-world.md",
            "public",
            "site",
            "hello world",
            Some("thisisaslug"),
        );
        insta::assert_yaml_snapshot!(path);

        let path = out_path(
            "_content/posts/hello-world.md",
            "public",
            ".",
            "hello world",
            None,
        );
        insta::assert_yaml_snapshot!(path);

        let path = out_path("hello-world.md", "public", ".", "hello world", None);
        insta::assert_yaml_snapshot!(path);

        let path = out_path(
            "site/_content/series/hello-world/index.md",
            "public",
            "site",
            "this is a series",
            None,
        );
        insta::assert_yaml_snapshot!(path);

        let path = out_path(
            "site/_content/series/hello-world/part-1.md",
            "public",
            "site",
            "Part One",
            None,
        );
        insta::assert_yaml_snapshot!(path);

        let path = out_path("site/_content/index.md", "public", "site", "", None);
        insta::assert_yaml_snapshot!(path);
    }
}
