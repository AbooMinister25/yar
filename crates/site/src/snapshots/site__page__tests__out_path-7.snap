---
source: crates/site/src/page.rs
expression: path
---
public/index.html
